driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -85,10 +91,46 @@ public SessionContext getSessionContext() {
 
     @Override
     public ConnectionSource getWriteConnectionSource() {
-        ConnectionSource writeConnectionSource = wrapped.getWriteConnectionSource();
+        ConnectionSource writeConnectionSource = getWrappedWriteConnectionSource();
         return new SessionBindingConnectionSource(writeConnectionSource);
     }
 
+    private ConnectionSource getWrappedWriteConnectionSource() {
+        ConnectionSource connectionSource = wrapped.getWriteConnectionSource();
+        if (isActiveShardedTxn()) {
+            SingleServerBinding binding = getSingleServerBinding();
+            connectionSource = binding.getWriteConnectionSource();
+            binding.release();
+        }
+        return connectionSource;
+    }
+
+    private ConnectionSource getWrappedReadConnectionSource() {
+        ConnectionSource connectionSource = wrapped.getReadConnectionSource();
+        if (isActiveShardedTxn()) {
+            SingleServerBinding binding = getSingleServerBinding();
+            connectionSource = binding.getReadConnectionSource();
+            binding.release();
+        }
+        return connectionSource;
+    }
+
+    private boolean isActiveShardedTxn() {
+        return session.hasActiveTransaction() && wrapped.getCluster().getDescription().getType() == ClusterType.SHARDED;
+    }
+
+    private void setPinnedMongosAddress() {
+        if (session.getPinnedMongosAddress() == null) {
+            Server server = wrapped.getCluster().selectServer(new ReadPreferenceServerSelector(wrapped.getReadPreference()));
+            session.setPinnedMongosAddress(server.getDescription().getAddress());
+        }
+    }
+
+    private SingleServerBinding getSingleServerBinding() {
```

Comment:
Once you DRY up the code above, I'd inline this method.  I find it a bit confusing as is because the most important part of it -- that it's pinning the mongos -- is not clear from the method name.

Reply:
I have inlined the method.